xds/src/main/java/io/grpc/xds/LeastRequestLoadBalancer.java

```diff
@@ -0,0 +1,421 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.ConnectivityState.CONNECTING;
+import static io.grpc.ConnectivityState.IDLE;
+import static io.grpc.ConnectivityState.READY;
+import static io.grpc.ConnectivityState.SHUTDOWN;
+import static io.grpc.ConnectivityState.TRANSIENT_FAILURE;
+import static io.grpc.xds.LeastRequestLoadBalancerProvider.DEFAULT_CHOICE_COUNT;
+import static io.grpc.xds.LeastRequestLoadBalancerProvider.MAX_CHOICE_COUNT;
+import static io.grpc.xds.LeastRequestLoadBalancerProvider.MIN_CHOICE_COUNT;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+import io.grpc.Attributes;
+import io.grpc.ClientStreamTracer;
+import io.grpc.ClientStreamTracer.StreamInfo;
+import io.grpc.ConnectivityState;
+import io.grpc.ConnectivityStateInfo;
+import io.grpc.EquivalentAddressGroup;
+import io.grpc.LoadBalancer;
+import io.grpc.Metadata;
+import io.grpc.Status;
+import io.grpc.xds.ThreadSafeRandom.ThreadSafeRandomImpl;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.Nonnull;
+
+/**
+ * A {@link LoadBalancer} that provides least request load balancing based on
+ * outstanding request counters.
+ * It works by sampling a number of subchannels and picking the one with the
+ * fewest amount of outstanding requests.
+ * The default sampling amount of two is also known as
+ * the "power of two choices" (P2C).
+ */
+final class LeastRequestLoadBalancer extends LoadBalancer {
+  @VisibleForTesting
+  static final Attributes.Key<Ref<ConnectivityStateInfo>> STATE_INFO =
+      Attributes.Key.create("state-info");
+  @VisibleForTesting
+  static final Attributes.Key<AtomicInteger> IN_FLIGHTS =
+      Attributes.Key.create("in-flights");
+
+  private final Helper helper;
+  private final ThreadSafeRandom random;
+  private final Map<EquivalentAddressGroup, Subchannel> subchannels =
+      new HashMap<>();
+
+  private ConnectivityState currentState;
+  private LeastRequestPicker currentPicker = new EmptyPicker(EMPTY_OK);
+  private int choiceCount = DEFAULT_CHOICE_COUNT;
+
+  LeastRequestLoadBalancer(Helper helper) {
+    this(helper, ThreadSafeRandomImpl.instance);
+  }
+
+  LeastRequestLoadBalancer(Helper helper, ThreadSafeRandom random) {
```

Comment:
`private`? Why introduce yet a 2nd constructor? For testing?

Reply:
Woops.. Nice catch!
This is an artifact of me using a combination of RoundRobinLoadBalancer and ClusterImplLoadBalancer for inspiration when making this.
This constructor should have been used in the tests similarly to the tests for ClusterImplLoadBalancer but wasn't.
As a result of using this constructor in the tests I realized a few things that I also have fixed now:

* There was a small bug in `ReadyPicker.isEquivalentTo` that would have prevented runtime updates to choiceCount unless the aggregated state also happened to change.
* There were no tests that actually verified that `updateBalancingState` was being called with a correct `ReadyPicker`.
* I realized I have no idea why I implemented a `FakeRandom` and didn't just use a normal mock instead.